compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Patterns.cpp

```diff
@@ -811,20 +811,141 @@ struct CmdDispatchOpPattern
     return success();
   }
 
+  // Returns the fully qualified export name (@executable::@variant::@export).
+  SymbolRefAttr getExportRef(IREE::HAL::ExecutableExportOp exportOp) const {
+    auto variantOp =
+        exportOp->getParentOfType<IREE::HAL::ExecutableVariantOp>();
+    auto executableOp = variantOp->getParentOfType<IREE::HAL::ExecutableOp>();
+    return SymbolRefAttr::get(executableOp.getSymNameAttr(),
+                              {
+                                  FlatSymbolRefAttr::get(variantOp),
+                                  FlatSymbolRefAttr::get(exportOp),
+                              });
+  }
+
+  // Selects the ordinal for the given |baseExportOp| and calculates its
+  // workgroup count. The ordinal may be different than the ordinal of the
+  // export itself if any fallbacks are specified. Each export condition will be
+  // evaluated and the first that matches will be returned.
+  //
+  // As an example, a fallback chain of @0 -> @1 -> @2 (with @0 being the
+  // highest priority) would result in a decision tree:
+  //   %ordinal, %workgroups = scf.if %cond0 {
+  //     %ordinal0 = hal.executable.export.ordinal @0
+  //     %workgroups0 = calculate for @0
+  //     scf.yield %ordinal0, %workgroups0
+  //   } else {
+  //     %ordinal12, %workgroups12 = scf.if %cond1 {
+  //       %ordinal1 = hal.executable.export.ordinal @1
+  //       %workgroups1 = calculate for @1
+  //       scf.yield %ordinal1, %workgroups1
+  //     } else {
+  //       %ordinal2 = hal.executable.export.ordinal @2
+  //       %workgroups2 = calculate for @2
+  //       scf.yield %ordinal2, %workgroups2
+  //     }
+  //     scf.yield %ordinal12, %workgroups12
+  //   }
+  std::tuple<Value, std::array<Value, 3>>
+  selectExport(Location loc, IREE::HAL::ExecutableExportOp baseExportOp,
+               Value device, ValueRange workload, OpBuilder &builder) const {
+    if (!baseExportOp.getConditionBody()) {
+      // No fallback - fast path to just the base export.
+      Value ordinal = builder.create<IREE::HAL::ExecutableExportOrdinalOp>(
+          loc, builder.getIndexType(), getExportRef(baseExportOp));
+      auto workgroupCount =
+          baseExportOp.calculateWorkgroupCount(loc, device, workload, builder);
+      return std::make_tuple(ordinal, workgroupCount);
```

Comment:
There was a recent discussion on discord with regards to the use of `make_pair`'s type deduction, and the same logic applies to `make_tuple`: https://discord.com/channels/689900678990135345/689906000043573354/1362622689176719531

nit: there is a preference for just the following
```suggestion
      return {ordinal, workgroupCount};
```